### *“Success is no accident. It is hard work, perseverance, learning, studying, sacrifice and most of all, love of what you are doing or learning to do.”*

*-- Pele, Brazilian soccer player*

You're doing well and that's not  by accident.  So continue to work hard!  Keep up the good work!

----

### YOUR NAME HERE:  Brandon Rank

#### List the names of anyone you helped here: Mo Asad, Luke Faro

#### List the names of anyone who helped you here: Lucas Acosta-Morales, Mo Asad
----

# HW SQL 2: Using Aggregate Functions with Group By

**Be sure to use your SQL cheat sheet to help you in this homework!**

The goal of this homework will be to practice selecting columns, filtering rows, using these concepts.

* Perform arithmetic operations on columns using  * / + -
* Using these aggregate functions:
  * MIN/MAX/AVG/SUM/COUNT
* Integer Division
  * Remember:  4 / 3 = 1 in the world of SQL!
* Using AS to create Aliases
* GROUP BY
* HAVING
* ORDER BY

Recall that the correct order of a SQL query is below.

    """SELECT column_name(s)
        FROM table_name
        WHERE condition                     
        GROUP BY column_name(s)
        HAVING condition
        ORDER BY column_name(s);"""


## Create a SQLite database from a csv file

Remember: To use SQLite, we need to import some special packages and to create a database (db) file from a csv file.

### !!!!!!!!!!!!!!!!!!!!!!So please run the cells below!!!!!!!!!!!!!!!!!!!!

In [ ]:
#This cell imports the SQLite package and 2 other packages that allow us to manipulate tables of data.

import pandas as pd
import numpy as np
import sqlite3 as sql

In [ ]:
#Read in a csv with Hearts On Fire data
hearts_on_fire = pd.read_csv('HoF_2020_2021.csv')

#Make a connection to a database file ('hearts_on_fire.db').
#If the file does not yet exist, it will be created.
conn = sql.connect('hearts_on_fire.db')

#The to_sql method writes records stored in a DataFrame (table) to a SQLite database.
#It requires 2 inputs to run:
#     the name of TABLE inside the database file
#     and a connection to the database where the table lives.
#So essentially, the code below dumps the contents of the hearts_on_fire dataframe
#     into a table called weather in a hearts_on_fire.db file.

try:
    hearts_on_fire.to_sql('hearts_on_fire', conn, index = False)
except ValueError:
    print("""Dr. R Note: A ValueError occurred. That's probably fine and likely just means
           that you've run this cell twice and you're getting an error
           because the DB is already created.  :) """)

## Understanding the Data

Yet again, we will be using the ```hearts_on_fire``` data.  

Recall that Hearts on Fire is a campaign that DSU runs every year to fundraise donations.

Here's what each of the columns mean.

* donation:  the amount the contributor contributed
* state:  where the alumnus lives
* affiliation: the status of the donor at the time of the donation
  * Indicated if the contributor was an alumni, student, family member, etc., when the donation was made.
* class_year:  the year the person graduated.
  * **Notice that some years are NaN - this is the python way to say NULL.**
* preferred_division:  where the contributor wishes to send their funds.
  * **Notice that some years are None - this is another python way to say NULL.**
* date_submitted: the date and time of the contributor made the donation
* donation year: The year of the hearts on fire campaign

In [ ]:
#Create the sql statement between the triple quotes.
sql_statement = """SELECT *
                   FROM hearts_on_fire;"""

#Run the query, passing it and the connection (conn) into the needed method.
#You get a dataframe with the results of the query.
query_results = pd.read_sql_query(sql_statement, conn)

#Print the dataframe results.
query_results

donation state affiliation  class_year   preferred_division  \
0         50.0    CA      Alumni      1997.0      Performing Arts   
1         37.0    PA     Student      2020.0             Business   
2         25.0    PA      Alumni         NaN                 None   
3         10.0    PA     Student      2022.0         Science/Math   
4         50.0    PA      Family         NaN                 None   
...        ...   ...         ...         ...                  ...   
1779     100.0    NY      Alumni      2019.0  Liberal Arts/Social   
1780      20.0    FL      Family      2023.0                 None   
1781     100.0    WV      Alumni      1970.0         Science/Math   
1782     250.0    DE      Alumni      2007.0  Liberal Arts/Social   
1783    2500.0    CT      Alumni      1971.0                 None   

       date_submitted  donation_year  
0     1/29/2020 10:57           2020  
1     2/12/2020 21:31           2020  
2     2/10/2020 15:10           2020  
3     2/12/2020 17:53           2020  
4     2/12/2020 15:47           2020  
...               ...            ...  
1779   2/8/2021 11:53           2021  
1780   2/9/2021 11:44           2021  
1781  2/10/2021 15:01           2021  
1782  2/10/2021 18:23           2021  
1783  2/10/2021 13:09           2021  

[1784 rows x 7 columns]

---

# **RECALL:**  Points will be deducted for not adhering to conventions.  

So remember to capitalize SQL keywords and  to line up the keywords in a column.

---

# QUESTION 1

When performing data analysis, you should be very aware of which columns have missing values.  Doing so will help you be careful about using COUNT(\*) or COUNT(col_name) in appropriate places.

Generate a table showing how many nulls there are per column.  

Rename the columns as shown.

So your final answer should match the table below, exactly.

| missing_donation | missing_state | missing_affiliation | missing_class_year | missing_preferred_division | missing_date_submitted | missing_donation_year |
|:-----------------:|:--------------:|:-------------------:|:------------------:|:--------------------------:|:----------------------:|:---------------------:|
|         0         |        3       |          0          |         683        |             936            |            0           |           0           |
|                   |                |                     |                    |                            |                        |                       |

In [ ]:
# CodeGrade Tag: question_1'

sql_statement1 = """SELECT COUNT(*)-COUNT(donation) AS missing_donation, COUNT(*)-COUNT(state) AS missing_state, COUNT(*)-COUNT(affiliation) AS missing_affiliation, COUNT(*)-COUNT(class_year) AS missing_class_year, COUNT(*)-COUNT(preferred_division) AS missing_preferred_division, COUNT(*)-COUNT(date_submitted) AS missing_date_submitted, COUNT(*)-COUNT(donation_year) AS missing_donation_year
                    FROM hearts_on_fire;"""

query_results1 = pd.read_sql_query(sql_statement1, conn)
query_results1

missing_donation  missing_state  missing_affiliation  missing_class_year  \
0                 0              3                    0                 683   

   missing_preferred_division  missing_date_submitted  missing_donation_year  
0                         936                       0                      0

# QUESTION 2

In one query, show the year and the largest/smallest/mean donation amount for that year.


In [ ]:
# CodeGrade Tag: question_2'

sql_statement2 = """SELECT donation_year, MAX(donation), MIN(donation), AVG(donation)
                    FROM hearts_on_fire
                    GROUP BY donation_year;"""

query_results2 = pd.read_sql_query(sql_statement2, conn)
query_results2

donation_year  MAX(donation)  MIN(donation)  AVG(donation)
0           2020        10000.0            1.0     139.423485
1           2021        10000.0            5.0     176.956835

# QUESTION 3

Determine how many contributors there were per state.

Order your result from the state with the highest number of contributors to the lowest.  

Rename the column with the number of contributors to appear as ``num_contributors``.



In [ ]:
# CodeGrade Tag: question_3
sql_statement3 = """SELECT state, COUNT(state) AS num_contributors
                    FROM hearts_on_fire
                    GROUP BY state
                    ORDER BY num_contributors DESC;"""


query_results3 = pd.read_sql_query(sql_statement3, conn)
query_results3

state  num_contributors
0     PA              1199
1     NJ               193
2     NY               104
3     DE                47
4     MD                39
5     CA                27
6     CT                16
7     FL                16
8     VA                14
9     NC                12
10    OH                12
11    CO                11
12    IL                11
13    MA                 8
14    TX                 8
15    MI                 7
16    DC                 5
17    NH                 5
18    SC                 5
19    TN                 5
20    GA                 4
21    NV                 3
22    OR                 3
23    RI                 3
24    WA                 3
25    AZ                 2
26    IN                 2
27    ME                 2
28    MO                 2
29    NM                 2
30    OK                 2
31    WV                 2
32    AK                 1
33    HI                 1
34    KY                 1
35    LA                 1
36    MN                 1
37    MT                 1
38    UT                 1
39  None                 0

# QUESTION 4

That was a lot of rows!  Redo the previous question so that only the top 10 states (in terms of number of contributors) show.  

HINT:  Your answer should not have a where clause or a having clause. There is a function that limits the number of rows returned.


In [ ]:
# CodeGrade Tag: question_4
sql_statement4 = """SELECT state, COUNT(state) AS num_contributors
                    FROM hearts_on_fire
                    GROUP BY state
                    ORDER BY num_contributors DESC
                    LIMIT 10;"""

query_results4 = pd.read_sql_query(sql_statement4, conn)
query_results4

state  num_contributors
0    PA              1199
1    NJ               193
2    NY               104
3    DE                47
4    MD                39
5    CA                27
6    CT                16
7    FL                16
8    VA                14
9    NC                12

# QUESTION 5

Recall that the Northeast Region is CT, ME, MA, NH, RI, VT, NJ, NY, PA.
  
Write a query to determine which states OUTSIDE of the Northeast Region had the 3 largest amounts of alumni contributors in the year 2021.

Your columns should be pertinent to the question at hand and your final table should only have 3 rows, making your answer clear.


In [ ]:
# CodeGrade Tag: question_5
sql_statement5 = """SELECT state, COUNT(affiliation) AS num_contributors, donation_year
                    FROM hearts_on_fire
                    WHERE (state NOT IN ('CT', 'ME', 'MA', 'NH', 'RI', 'VT', 'NJ', 'NY', 'PA')) AND (donation_year = 2021) AND (affiliation = 'Alumni')
                    GROUP BY state
                    ORDER BY COUNT(state) DESC
                    LIMIT 3;"""

query_results5 = pd.read_sql_query(sql_statement5, conn)
query_results5

state  num_contributors  donation_year
0    DE                17           2021
1    CA                10           2021
2    MD                 9           2021

# QUESTION 6

For the year 2020, show the number of donors per the amount of donation.  

In the final results, order the donations from smallest to largest.

In [ ]:
# CodeGrade Tag: question_6
sql_statement6 = """SELECT donation, COUNT(donation) AS num_donors
                    FROM hearts_on_fire
                    WHERE (donation_year = 2020)
                    GROUP BY donation
                    ORDER BY donation ASC;"""

query_results6 = pd.read_sql_query(sql_statement6, conn)
query_results6

donation  num_donors
0       1.00           3
1       2.00           3
2       3.00           1
3       5.00          45
4       5.98           1
5       6.00           1
6      10.00          62
7      15.00          15
8      20.00          55
9      20.08           1
10     20.16           1
11     20.17           1
12     25.00         185
13     30.00          24
14     35.00           2
15     36.00           1
16     37.00           2
17     40.00          10
18     45.00           1
19     50.00         124
20     55.00           1
21     60.00           2
22     67.00           1
23     72.00           1
24     75.00           7
25    100.00         169
26    100.19           1
27    120.00           1
28    125.00           4
29    150.00          21
30    200.00          15
31    225.00           1
32    250.00          29
33    300.00           7
34    400.00           3
35    500.00          32
36    750.00           1
37    800.00           1
38   1000.00          11
39   1500.00           1
40   2000.00           1
41   2500.00           2
42   3000.00           1
43   4348.50           1
44   5000.00           1
45  10000.00           2

# QUESTION 7

There were a lot of 1s in the prior table.  

Run the same query but this time, show only those rows where more than 1 person contributed the given amount.

In [ ]:
# CodeGrade Tag: question_7
sql_statement7 = """SELECT donation, COUNT(donation) AS num_donors
                    FROM hearts_on_fire
                    WHERE (donation_year = 2020)
                    GROUP BY donation
                    HAVING COUNT(donation) > 1
                    ORDER BY donation ASC;"""

query_results7 = pd.read_sql_query(sql_statement7, conn)
query_results7

donation  num_donors
0        1.0           3
1        2.0           3
2        5.0          45
3       10.0          62
4       15.0          15
5       20.0          55
6       25.0         185
7       30.0          24
8       35.0           2
9       37.0           2
10      40.0          10
11      50.0         124
12      60.0           2
13      75.0           7
14     100.0         169
15     125.0           4
16     150.0          21
17     200.0          15
18     250.0          29
19     300.0           7
20     400.0           3
21     500.0          32
22    1000.0          11
23    2500.0           2
24   10000.0           2

# QUESTION 8

Show how many contributors there were per donation year/affiliation.  

Order your table so that the donation years increase while the affiliations are in alphabetical order per year.

In [ ]:
# CodeGrade Tag: question_8
sql_statement8 = """SELECT donation_year, affiliation, COUNT(affiliation) AS num_contributors
                    FROM hearts_on_fire
                    GROUP BY donation_year, affiliation
                    ORDER BY donation_year ASC, affiliation ASC"""

query_results8 = pd.read_sql_query(sql_statement8, conn)
query_results8

donation_year affiliation  num_contributors
0           2020      Alumni               470
1           2020    Employee               175
2           2020      Family               128
3           2020      Friend                31
4           2020     Student                51
5           2021      Alumni               466
6           2021    Employee               167
7           2021      Family               150
8           2021      Friend                61
9           2021     Student                85

# QUESTION 9

Show the average donations per alumni class year, ordered from highest average to lowest.

Name the average column:  ```avg_donation```.

In [ ]:
# CodeGrade Tag: question_9
sql_statement9 = """SELECT class_year, AVG(donation) AS avg_donation
                    FROM hearts_on_fire
                    GROUP BY class_year
                    ORDER BY avg_donation DESC;"""

query_results9 = pd.read_sql_query(sql_statement9, conn)
query_results9

class_year  avg_donation
0       1994.0   1139.090909
1       1971.0   1129.545455
2       1985.0    899.230769
3       1969.0    788.690476
4       1990.0    746.153846
5       1970.0    576.562500
6       1997.0    538.636364
7       1979.0    530.000000
8       1984.0    335.000000
9       1977.0    300.000000
10      1976.0    275.000000
11      1981.0    237.500000
12      2007.0    220.685185
13      1978.0    216.666667
14      2000.0    216.428571
15      1982.0    208.571429
16      1980.0    195.000000
17      1983.0    185.000000
18      1998.0    181.491071
19      1987.0    167.857143
20      1991.0    165.000000
21      1996.0    156.111111
22      1975.0    153.333333
23      1973.0    140.666667
24      2003.0    129.166667
25         NaN    127.705959
26      1993.0    125.000000
27      2009.0    123.500000
28      1988.0    115.625000
29      1986.0    113.333333
30      2002.0    110.333333
31      2012.0    102.777778
32      2004.0    100.227273
33      1974.0    100.000000
34      2006.0     91.750000
35      1995.0     89.125000
36      2005.0     86.739130
37      2014.0     83.888889
38      1992.0     83.333333
39      2001.0     81.250000
40      2023.0     76.921053
41      2019.0     72.265625
42      1999.0     69.783784
43      2015.0     65.441176
44      2016.0     65.222500
45      2010.0     64.341250
46      1972.0     62.500000
47      2017.0     62.068966
48      2020.0     58.843243
49      1989.0     58.333333
50      2013.0     55.666667
51      2008.0     54.226667
52      2011.0     52.407407
53      2021.0     51.296296
54      2025.0     50.666667
55      2022.0     50.000000
56      2024.0     43.804348
57      2018.0     31.621622

# QUESTION 10

Notice that in your previous question, there was a category for NaN where the average was about 127.71.

Write an english sentence for what you learn from this row.  The value $127.71 should be in your answer.

From this row, we learn that for the contributors who did not disclose their class_year when submitting their donation, the average contribution given was approximately $127.71.

# QUESTION 11

We'll say a high-donor is one who contributes 500 or more dollars.  

Show the number of high-donors in each state.

Order your results from the largest number of high donors to the smallest.

Also, name the column with the number of high donors as: ``num_high_donors``.


In [ ]:
# CodeGrade Tag: question_11
sql_statement11 = """SELECT state, COUNT(*) AS num_high_donors
                     FROM hearts_on_fire
                     WHERE donation >= 500
                     GROUP BY state
                     ORDER BY COUNT(state) DESC;"""

query_results11 = pd.read_sql_query(sql_statement11, conn)
query_results11

state  num_high_donors
0     PA               75
1     NJ               21
2     DE                6
3     MD                5
4     MA                3
5     CA                2
6     CT                2
7     MI                2
8     MO                2
9     NY                2
10    VA                2
11    CO                1
12    IL                1
13    ME                1
14    NC                1
15    NM                1
16    OH                1
17    WA                1
18  None                1

# QUESTION 12

Ask an interesting question that was not already asked, one that uses GROUPBY.  

Make it interesting and informative, ie, don't give me a lame question to just get points.

What is the total amount of donations from students in the class of 2021 within the Northeast Region per state, ordered from the lowest total to the highest.

In [ ]:
# CodeGrade Tag: question_12

sql_statement12 = """SELECT state, SUM(donation) AS total_donations
                     FROM hearts_on_fire
                     WHERE (state IN ('CT', 'ME', 'MA', 'NH', 'RI', 'VT', 'NJ', 'NY', 'PA')) AND (affiliation = 'Student') AND (class_year = 2021)
                     GROUP BY state
                     ORDER BY total_donations ASC;"""

query_results12 = pd.read_sql_query(sql_statement12, conn)
query_results12

state  total_donations
0    NY             25.0
1    NJ            125.0
2    PA            220.0

# QUESTION 13

The following query fails in DataCamp.  

``` SELECT country, AVG(gross) ```

```FROM films```

```WHERE AVG(gross) > 1000000```

```GROUP BY country```


### Why?-  What's Wrong?

The query will fail in DataCamp due to the fact that it is trying to run a query that contains a WHERE function which contains an aggregate function. When using WHERE, you cannot use aggregate functions. To fix the error, we could instead use the HAVING function after GROUP BY. This would be HAVING AVG(gross) > 1000000. The HAVING function is what we must use whenever we wish to utilize an aggregate function within a WHERE search.

# QUESTION 14

Did you follow the conventions expected?  For this question I will give/takeaway points for doing so.

----

## Closing the DB connection.

Throughout this whole file, we've kept our connection to the Database (the hearts_on_fire.db file) open.  

We must close the connection when done.

Run the following cell when you are done working on this HW.


In [ ]:
#Close the cursor and connection
conn.close()

---
All done! Good job!